DNN Platform/Library/Services/Users/PurgeDeletedUsers.cs

```diff
@@ -0,0 +1,94 @@
+// 
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
+// documentation files (the "Software"), to deal in the Software without restriction, including without limitation 
+// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and 
+// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+// 
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions 
+// of the Software.
+// **********************************
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED 
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL 
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
+// CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
+// DEALINGS IN THE SOFTWARE.
+#endregion
+#region Usings
+
+using DotNetNuke.Entities.Portals;
+using DotNetNuke.Entities.Users;
+using DotNetNuke.Services.Scheduling;
+using System;
+
+#endregion
+
+
+namespace DotNetNuke.Services.Users
+{
+    public class PurgeDeletedUsers : SchedulerClient
+    {
+
+        public PurgeDeletedUsers(ScheduleHistoryItem objScheduleHistoryItem)
+        {
+            ScheduleHistoryItem = objScheduleHistoryItem;
+        }
+
+        public override void DoWork()
+        {
+            try
+            {
+                foreach (PortalInfo portal in new PortalController().GetPortals())
+                {
+                    var settings = new PortalSettings(portal.PortalID);
+                    if (settings.DataConsentActive)
+                    {
+                        if (settings.DataConsentUserDeleteAction == PortalSettings.UserDeleteAction.DelayedHardDelete)
+                        {
+                            var thresholdDate = DateTime.Now;
+                            switch (settings.DataConsentDelayMeasurement)
+                            {
+                                case "h":
+                                    thresholdDate = DateTime.Now.AddHours(-1 * settings.DataConsentDelay);
+                                    break;
+                                case "d":
+                                    thresholdDate = DateTime.Now.AddDays(-1 * settings.DataConsentDelay);
+                                    break;
+                                case "w":
+                                    thresholdDate = DateTime.Now.AddDays(-7 * settings.DataConsentDelay);
+                                    break;
+                            }
+                            var deletedUsers = UserController.GetDeletedUsers(portal.PortalID);
+                            foreach (UserInfo user in deletedUsers)
```

Comment:
Do we have any concerns on the amount of records that we will get with this?  Going forward it should be a small number, but not sure if this could get out of hand and limit the ability to process successfully with the 30 second query timeout.

Reply:
It runs as a scheduled task. I could not imagine a scenario where the amount of soft deleted accounts was too large to handle in memory or time. Isn't it OK for this to time out and pick up again after 30 mins in the exceptional case that the site has massive amounts of soft deleted users?

Reply:
if a sproc timed out due to the amount of users, no records will be affected ad the same happen again on next run. However, I agree, a huge number should be an edge case.

Reply:
> 
> 
> This looks really great, and thank you for the contribution! It's going to make a lot of people quite happy to see this in the release notes for v9.4.0!
> 
> Can you easily move the sql changes into the main 940 file? This early in the cycle I'd rather keep them in a single file, but since we really don't have anything defined for sql changes it's really just my personal preference for release management and upgrade performance and not a required change at all.

Done